File: src/main/java/net/skinsrestorer/mojangapi/CacheManager.java
package net.skinsrestorer.mojangapi;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Mono;

import javax.annotation.Nullable;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

@Slf4j
public class CacheManager implements AutoCloseable {
  private final Cache<String, DatabaseResult<UUID>> nameToUUIDCache = Caffeine.newBuilder()
    .expireAfterWrite(6, TimeUnit.HOURS)
    .build();
  private final Cache<UUID, DatabaseResult<SkinProperty>> uuidToSkinCache = Caffeine.newBuilder()
    .expireAfterWrite(6, TimeUnit.HOURS)
    .build();

  public void putNameToUUID(String name, @Nullable UUID uuid, LocalDateTime createdAt) {
    nameToUUIDCache.put(name, new DatabaseResult<>(createdAt.toEpochSecond(ZoneOffset.UTC), uuid));
  }

  public Mono<DatabaseResult<UUID>> getNameToUUID(String name) {
    return Mono.justOrEmpty(nameToUUIDCache.getIfPresent(name));
  }

  public void putUUIDToSkin(UUID uuid, @Nullable SkinProperty skinProperty, LocalDateTime createdAt) {
    uuidToSkinCache.put(uuid, new DatabaseResult<>(createdAt.toEpochSecond(ZoneOffset.UTC), skinProperty));
  }

  public Mono<DatabaseResult<SkinProperty>> getUUIDToSkin(UUID uuid) {
    return Mono.justOrEmpty(uuidToSkinCache.getIfPresent(uuid));
  }

  @Override
  public void close() {
    nameToUUIDCache.invalidateAll();
    uuidToSkinCache.invalidateAll();
    log.info("CacheManager closed and all caches invalidated.");
  }

  public record DatabaseResult<T>(long createdAt, @Nullable T value) {
  }

  public record SkinProperty(
    String value,
    String signature
  ) {
  }
}

File: src/cache-manager.ts
import { LRUCache } from "lru-cache";

export interface DatabaseResult<T> {
  createdAt: number;
  value: T | null;
}

export interface SkinProperty {
  value: string;
  signature: string;
}

export interface CacheManager {
  putNameToUUID: (name: string, uuid: string | null, createdAt: Date) => void;
  getNameToUUID: (name: string) => Promise<DatabaseResult<string> | undefined>;
  putUUIDToSkin: (
    uuid: string,
    skinProperty: SkinProperty | null,
    createdAt: Date,
  ) => void;
  getUUIDToSkin: (
    uuid: string,
  ) => Promise<DatabaseResult<SkinProperty> | undefined>;
  close: () => void;
}

export function createCacheManager(): CacheManager {
  // Create caches with similar settings to the Java implementation
  const nameToUUIDCache = new LRUCache<string, DatabaseResult<string>>({
    max: 10000, // Maximum size of 10,000 entries
    ttl: 1000 * 60 * 60 * 6, // 6 hours (matching Java implementation)
  });

  const uuidToSkinCache = new LRUCache<string, DatabaseResult<SkinProperty>>({
    max: 10000, // Maximum size of 10,000 entries
    ttl: 1000 * 60 * 60 * 6, // 6 hours (matching Java implementation)
  });

  return {
    putNameToUUID(name: string, uuid: string | null, createdAt: Date) {
      nameToUUIDCache.set(name.toLowerCase(), {
        createdAt: Math.floor(createdAt.getTime() / 1000),
        value: uuid,
      });
    },

    async getNameToUUID(
      name: string,
    ): Promise<DatabaseResult<string> | undefined> {
      return nameToUUIDCache.get(name.toLowerCase());
    },

    putUUIDToSkin(
      uuid: string,
      skinProperty: SkinProperty | null,
      createdAt: Date,
    ) {
      uuidToSkinCache.set(uuid.toLowerCase(), {
        createdAt: Math.floor(createdAt.getTime() / 1000),
        value: skinProperty,
      });
    },

    async getUUIDToSkin(
      uuid: string,
    ): Promise<DatabaseResult<SkinProperty> | undefined> {
      return uuidToSkinCache.get(uuid.toLowerCase());
    },

    close() {
      nameToUUIDCache.clear();
      uuidToSkinCache.clear();
      console.log("CacheManager closed and all caches invalidated.");
    },
  };
}
